File: 0206-reverse-linked-list/0206-reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        if(head == null) return head;
        return reverseList(null, head, head.next);
    }

    private ListNode reverseList(ListNode prev, ListNode curr, ListNode next) {
        curr.next = prev;
        prev = curr;
        curr = next;

        if(next == null)
            return prev;
        
        next = next.next;
        return reverseList(prev, curr, next);
    }
}

File: 0206-reverse-linked-list/0206-reverse-linked-list.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun reverseList(head: ListNode?): ListNode? {
        var (prev, curr, next) = Triple<ListNode?, ListNode?, ListNode?>(null, head, head?.next)

        if(curr == null)
            return head

        while(curr != null) {
            curr.next = prev
            prev = curr
            curr = next
            next = next?.next
        }

        return prev
    }
}